File: orchestrator/src/queue/queue.ts
import crypto from 'node:crypto'
import type { Db } from '../db/index.js'

export type JobStatus = 'pending' | 'claimed' | 'completed' | 'failed'

export interface Job<P = Record<string, unknown>> {
  id: string
  jobType: string
  payload: P
  status: JobStatus
  priority: number
  attempts: number
  maxAttempts: number
  error?: Record<string, unknown>
  createdAt: string
  claimedAt?: string
  completedAt?: string
}

export interface EnqueueInput {
  jobType: string
  payload?: Record<string, unknown>
  priority?: number
  maxAttempts?: number
  id?: string
}

interface JobRow {
  id: string
  job_type: string
  payload_json: string
  status: string
  priority: number
  attempts: number
  max_attempts: number
  error_json: string | null
  created_at: string
  claimed_at: string | null
  completed_at: string | null
}

function rowToJob(row: JobRow): Job {
  return {
    id: row.id,
    jobType: row.job_type,
    payload: JSON.parse(row.payload_json) as Record<string, unknown>,
    status: row.status as JobStatus,
    priority: row.priority,
    attempts: row.attempts,
    maxAttempts: row.max_attempts,
    ...(row.error_json ? { error: JSON.parse(row.error_json) as Record<string, unknown> } : {}),
    createdAt: row.created_at,
    ...(row.claimed_at ? { claimedAt: row.claimed_at } : {}),
    ...(row.completed_at ? { completedAt: row.completed_at } : {}),
  }
}

export class JobQueue {
  constructor(private readonly db: Db) {}

  enqueue(input: EnqueueInput): Job {
    const id = input.id ?? crypto.randomUUID()
    const now = new Date().toISOString()
    this.db.prepare(`
      INSERT INTO queue_jobs (id, job_type, payload_json, priority, max_attempts, created_at)
      VALUES (?, ?, ?, ?, ?, ?)
    `).run(
      id,
      input.jobType,
      JSON.stringify(input.payload ?? {}),
      input.priority ?? 0,
      input.maxAttempts ?? 3,
      now,
    )
    return this.getJob(id)!
  }

  /** Atomically claim the next pending job. Returns null when queue is empty. */
  dequeue(jobType?: string): Job | null {
    const typeFilter = jobType ? `AND job_type = '${jobType.replace(/'/g, "''")}'` : ''
    const now = new Date().toISOString()

    // SQLite doesn't support RETURNING in older versions; use a transaction to claim atomically.
    const job = this.db.transaction(() => {
      const row = this.db.prepare<[], JobRow>(`
        SELECT * FROM queue_jobs
        WHERE status = 'pending' ${typeFilter}
        ORDER BY priority DESC, created_at ASC
        LIMIT 1
      `).get()

      if (!row) return null

      this.db.prepare(`
        UPDATE queue_jobs
        SET status = 'claimed', claimed_at = ?, attempts = attempts + 1
        WHERE id = ?
      `).run(now, row.id)

      return this.db.prepare<[string], JobRow>('SELECT * FROM queue_jobs WHERE id = ?').get(row.id) ?? null
    })()

    return job ? rowToJob(job) : null
  }

  complete(id: string): void {
    const now = new Date().toISOString()
    this.db.prepare(`
      UPDATE queue_jobs SET status = 'completed', completed_at = ? WHERE id = ?
    `).run(now, id)
  }

  fail(id: string, error: Record<string, unknown>): void {
    const now = new Date().toISOString()
    // If attempts < max_attempts, reset to pending for retry; otherwise mark failed.
    this.db.transaction(() => {
      const row = this.db.prepare<[string], { attempts: number; max_attempts: number }>(
        'SELECT attempts, max_attempts FROM queue_jobs WHERE id = ?',
      ).get(id)

      if (!row) return

      const nextStatus = row.attempts < row.max_attempts ? 'pending' : 'failed'
      this.db.prepare(`
        UPDATE queue_jobs
        SET status = ?, error_json = ?, completed_at = CASE WHEN ? = 'failed' THEN ? ELSE NULL END
        WHERE id = ?
      `).run(nextStatus, JSON.stringify(error), nextStatus, now, id)
    })()
  }

  /** Reclaim jobs that were claimed but never completed (process crash recovery). */
  reclaim(timeoutMs = 60_000): number {
    const cutoff = new Date(Date.now() - timeoutMs).toISOString()
    const result = this.db.prepare(`
      UPDATE queue_jobs
      SET status = 'pending', claimed_at = NULL
      WHERE status = 'claimed' AND claimed_at < ?
    `).run(cutoff)
    return result.changes
  }

  getJob(id: string): Job | null {
    const row = this.db.prepare<[string], JobRow>('SELECT * FROM queue_jobs WHERE id = ?').get(id)
    return row ? rowToJob(row) : null
  }

  listJobs(status?: JobStatus, limit = 50): Job[] {
    const where = status ? 'WHERE status = ?' : ''
    const params: unknown[] = status ? [status] : []
    const rows = this.db.prepare<unknown[], JobRow>(`
      SELECT * FROM queue_jobs ${where} ORDER BY created_at DESC LIMIT ${limit}
    `).all(...params)
    return rows.map(rowToJob)
  }

  pendingCount(jobType?: string): number {
    const where = jobType ? `AND job_type = ?` : ''
    const params: unknown[] = jobType ? [jobType] : []
    const row = this.db.prepare<unknown[], { n: number }>(
      `SELECT COUNT(*) AS n FROM queue_jobs WHERE status = 'pending' ${where}`,
    ).get(...params)
    return row?.n ?? 0
  }
}

File: packages/orchestrator-go/internal/queue/queue.go
package queue

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/ricardoborges-teachable/ai-setup/packages/orchestrator-go/internal/db"
)

// Status represents the lifecycle state of a job.
type Status string

const (
	StatusPending   Status = "pending"
	StatusClaimed   Status = "claimed"
	StatusCompleted Status = "completed"
	StatusFailed    Status = "failed"
)

// Job represents a queued background job.
type Job struct {
	ID           string         `json:"id"`
	JobType      string         `json:"jobType"`
	Payload      map[string]any `json:"payload"`
	Status       Status         `json:"status"`
	Priority     int            `json:"priority"`
	Attempts     int            `json:"attempts"`
	MaxAttempts  int            `json:"maxAttempts"`
	Error        map[string]any `json:"error,omitempty"`
	CreatedAt    string         `json:"createdAt"`
	ClaimedAt    string         `json:"claimedAt,omitempty"`
	CompletedAt  string         `json:"completedAt,omitempty"`
}

// EnqueueInput describes a job to enqueue.
type EnqueueInput struct {
	JobType     string
	Payload     map[string]any
	Priority    int
	MaxAttempts int
}

// Queue is a durable job queue backed by SQLite.
type Queue struct {
	db *db.DB
}

// New creates a new queue.
func New(database *db.DB) *Queue {
	return &Queue{db: database}
}

// Enqueue adds a job to the queue.
func (q *Queue) Enqueue(input EnqueueInput) (*Job, error) {
	id := uuid.NewString()
	now := time.Now().UTC().Format(time.RFC3339)
	payloadJSON, _ := json.Marshal(input.Payload)
	maxAttempts := input.MaxAttempts
	if maxAttempts <= 0 {
		maxAttempts = 3
	}

	_, err := q.db.Exec(
		`INSERT INTO queue_jobs (id, job_type, payload_json, priority, max_attempts, created_at) VALUES (?, ?, ?, ?, ?, ?)`,
		id, input.JobType, string(payloadJSON), input.Priority, maxAttempts, now)
	if err != nil {
		return nil, err
	}
	return q.Get(id)
}

// Dequeue atomically claims the next pending job.
func (q *Queue) Dequeue(jobType string) (*Job, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	var query string
	var args []any
	if jobType != "" {
		query = `SELECT id FROM queue_jobs WHERE status = 'pending' AND job_type = ? ORDER BY priority DESC, created_at ASC LIMIT 1`
		args = []any{jobType}
	} else {
		query = `SELECT id FROM queue_jobs WHERE status = 'pending' ORDER BY priority DESC, created_at ASC LIMIT 1`
		args = nil
	}

	var jobID string
	err := q.db.QueryRow(query, args...).Scan(&jobID)
	if err != nil {
		return nil, nil // empty queue
	}

	q.db.Exec(`UPDATE queue_jobs SET status = 'claimed', claimed_at = ?, attempts = attempts + 1 WHERE id = ?`, now, jobID)
	return q.Get(jobID)
}

// Complete marks a job as completed.
func (q *Queue) Complete(id string) error {
	now := time.Now().UTC().Format(time.RFC3339)
	_, err := q.db.Exec(`UPDATE queue_jobs SET status = 'completed', completed_at = ? WHERE id = ?`, now, id)
	return err
}

// Fail marks a job as failed; re-queues if attempts remain.
func (q *Queue) Fail(id string, errData map[string]any) error {
	errJSON, _ := json.Marshal(errData)
	now := time.Now().UTC().Format(time.RFC3339)

	var attempts, maxAttempts int
	q.db.QueryRow(`SELECT attempts, max_attempts FROM queue_jobs WHERE id = ?`, id).Scan(&attempts, &maxAttempts)

	var nextStatus string
	var completedAt string
	if attempts < maxAttempts {
		nextStatus = "pending"
	} else {
		nextStatus = "failed"
		completedAt = now
	}

	if completedAt != "" {
		_, err := q.db.Exec(
			`UPDATE queue_jobs SET status = ?, error_json = ?, completed_at = ? WHERE id = ?`,
			nextStatus, string(errJSON), completedAt, id)
		return err
	}
	_, err := q.db.Exec(
		`UPDATE queue_jobs SET status = ?, error_json = ?, claimed_at = NULL WHERE id = ?`,
		nextStatus, string(errJSON), id)
	return err
}

// Reclaim resets jobs claimed longer than timeoutMs ago.
func (q *Queue) Reclaim(timeoutMs int) (int, error) {
	cutoff := time.Now().Add(-time.Duration(timeoutMs) * time.Millisecond).UTC().Format(time.RFC3339)
	result, err := q.db.Exec(
		`UPDATE queue_jobs SET status = 'pending', claimed_at = NULL WHERE status = 'claimed' AND claimed_at < ?`,
		cutoff)
	if err != nil {
		return 0, err
	}
	n, _ := result.RowsAffected()
	return int(n), nil
}

// Get returns a job by ID.
func (q *Queue) Get(id string) (*Job, error) {
	var j Job
	var payloadJSON, errorJSON string

	err := q.db.QueryRow(
		`SELECT id, job_type, payload_json, status, priority, attempts, max_attempts, COALESCE(error_json,''), created_at, COALESCE(claimed_at,''), COALESCE(completed_at,'') FROM queue_jobs WHERE id = ?`, id).
		Scan(&j.ID, &j.JobType, &payloadJSON, &j.Status, &j.Priority, &j.Attempts, &j.MaxAttempts, &errorJSON, &j.CreatedAt, &j.ClaimedAt, &j.CompletedAt)
	if err != nil {
		return nil, fmt.Errorf("job not found: %s", id)
	}
	json.Unmarshal([]byte(payloadJSON), &j.Payload)
	if errorJSON != "" {
		json.Unmarshal([]byte(errorJSON), &j.Error)
	}
	return &j, nil
}

// List returns jobs optionally filtered by status.
func (q *Queue) List(status string, limit int) ([]Job, error) {
	var query string
	var args []any
	if status != "" {
		query = `SELECT id, job_type, payload_json, status, created_at FROM queue_jobs WHERE status = ? ORDER BY created_at DESC LIMIT ?`
		args = []any{status, limit}
	} else {
		query = `SELECT id, job_type, payload_json, status, created_at FROM queue_jobs ORDER BY created_at DESC LIMIT ?`
		args = []any{limit}
	}

	rows, err := q.db.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var jobs []Job
	for rows.Next() {
		var j Job
		var payloadJSON string
		if err := rows.Scan(&j.ID, &j.JobType, &payloadJSON, &j.Status, &j.CreatedAt); err != nil {
			continue
		}
		json.Unmarshal([]byte(payloadJSON), &j.Payload)
		jobs = append(jobs, j)
	}
	return jobs, rows.Err()
}
